### ISSUE TITLE
Calling .getType() does not work after manipulation

### ISSUE THREAD

**Issue by sfrooster** [dsherret/ts-simple-ast#41]
I have a ClassDeclaration, classDec, for the following class:
```typescript
        class MyClass {
            public x: number = 0;
        }
```
If I do something like:
```typescript
        classDec.getInstanceMembers()
            .filter(m => m.hasModifier(ts.SyntaxKind.PublicKeyword))
            .forEach(member => {
                if (member instanceof ParameterDeclaration || member instanceof PropertyDeclaration) {
                    let mType = member.getType();
                    ...
                }
            });
```
It blows-up because if something to do with an undefined symbol.

Here's the stack trace:
```
TypeError: Cannot read property 'flags' of undefined
    at getCheckFlags (/Users/mhaan/AppDev/tscompile/node_modules/ts-simple-ast/node_modules/typescript/lib/typescript.js:26031:26)
    at getTypeOfSymbol (/Users/mhaan/AppDev/tscompile/node_modules/ts-simple-ast/node_modules/typescript/lib/typescript.js:29477:17)
    at getTypeOfNode (/Users/mhaan/AppDev/tscompile/node_modules/ts-simple-ast/node_modules/typescript/lib/typescript.js:45506:24)
    at Object.getTypeAtLocation (/Users/mhaan/AppDev/tscompile/node_modules/ts-simple-ast/node_modules/typescript/lib/typescript.js:25507:31)
    at TypeChecker.getTypeAtLocation (/Users/mhaan/AppDev/tscompile/node_modules/ts-simple-ast/dist/compiler/tools/TypeChecker.js:51:62)
    at PropertyDeclaration.getType (/Users/mhaan/AppDev/tscompile/node_modules/ts-simple-ast/dist/compiler/base/TypedNode.js:8:50)
    at ast.getSourceFiles.forEach.processNS.forEach.classDec.getInstanceMembers.filter.forEach.member (/Users/mhaan/AppDev/tscompile/lib/app2.js:46:36)
    at Array.forEach (native)
```

Note that if I exclude PropertyDeclarations (by removing "|| member instanceof PropertyDeclaration") then that member is not found at all.

**Comment by sfrooster** [comment]
This appears to be related to multiple class declarations in the same file... ?

**Comment by sfrooster** [comment]
I think was due to violating the warning about using a changed AST - closing.

**Comment by dsherret** [comment]
Hmmm... that's strange. It should work with a changed AST. Are you sure this isn't a bug?

Are you talking about the warning on [this page](https://dsherret.github.io/ts-simple-ast/navigation/compiler-nodes)? That's about doing something like this:

```typescript
const classDec = sourceFile.getClass("MyClass")!;
let classCompilerNode = classDec.compilerNode;
classDec.setIsStatic(true);

// classCompilerNode is out of date now that the AST has been manipulated
classDec.getIsStatic(); // true
classCompilerNode.modifiers.some(m => m.getText() === "static"); // false, out of date
classCompilerNode = classDec.compilerNode; // update it with the new node
classCompilerNode.modifiers.some(m => m.getText() === "static"); // true
```

**Comment by dsherret** [comment]
I reproduced it:

```typescript
const ast = new TsSimpleAst();
const sourceFile = ast.addSourceFileFromText("myFile.ts", `function myFunction(param: string) {}`);
const param = sourceFile.getFunctions()[0].getParameters()[0];
expect(param.getType().getText()).to.equal("string");
param.setType("number");
expect(param.getType().getText()).to.equal("number");
```

Problem is that the internal type checker is not cleared or somehow updated after manipulating.

**Comment by sfrooster** [comment]
Yep, that's the warning.  What I was doing specifically was loading a SourceFile with the following code (in the SourceFile):

```typescript
export namespace Demo.UserProfiles.Commands {
    export class CreateUserProfile extends {
        public static readonly Version: string = "1.0";  

        constructor();
        constructor(Email: string, FirstName: string, LastName: string, SecurityToken: string);
        constructor(public Email: string = "", public FirstName?: string, public LastName?: string, public SecurityToken?: string) {
            super();
        }
    }

    export namespace Nested {
        export class CreateUserProfile2 {
            public static readonly Version: string = "1.0";

            public y: string = "xxx";

            constructor() { }
        }
    }
}
```

Then I use your library to process the namespaces and classes in the sourcefile.
Specifically, I:

- find the fully qualified classname and set it into a static member on the class - either set the value (initializer) if it's wrong, or create the whole thing if it's not there

- find all the instance members of the class and print some info out about them

With the file above, my code executes both bullet points on the first class, executes the first bullet point on the second class, and then blows-up as described executing the second bullet point on the second class.

I played around with it a bunch and found that as long as I made changes to only one class in the sourcefile per run, everything worked fine.

Now I've changed quite a few things around as part of other work (when I create a public static class member, it doesn't have the public keyword like I'd like, even though they're equivalent - I figured-out to just toggleModifier("public") or addModifier as long as I ignored types), so I'm not fully sure of  exactly how I had it.

One thing to note is that before processing the classes I process the namespaces in the sourcefile and capture the ClassDeclaration.  I use this captured ClassDeclaration later in processing the classes and maybe that's causing the issue - I shouldn't be holding on to those ClassDeclarations.

**Comment by sfrooster** [comment]
Just read your comment after posting mine - great!  I'm working on this now so let me know if I can pull a new version...

**Comment by dsherret** [comment]
@sfrooster unfortunately I have a busy few days ahead. I won't be able to get to it for a bit. What you can do for the time being is to call this hidden function that exists on any node after you're done manipulating and before calling `getType` again:

```typescript
(sourceFile as any).factory.resetProgram();
```

When I come up with the fix for this, you will be able to just remove all those statements. Sorry about that, but at least there's a workaround for the time being.

**Comment by sfrooster** [comment]
Ok, good to know.  I'll give that a shot.  I had been messing with some other workarounds, but I'll go with this for now if it works.

**Comment by dsherret** [comment]
Fixed in 0.38.1. You can remove all those `resetProgram` statements now.
